### PR TITLE
8310923: Refactor Currency tests to use JUnit

### PR DESCRIPTION
**Pull request by justin-curtis-lu**
Please review this PR which refactors Currency tests to use JUnit.

The most significant change occurs in `ValidateISO4217.java`. Other changes to this file excluding the JUnit refactoring include

- Tests are no longer dependent on each other (order of execution does not matter)
- Testing does not occur at the same time as data generation (The data is fully generated before any tests are executed)
- Various cleanup (dead-code, clarifying comments, more descriptive method and var names)

`Bug4512215.java` now renamed to `MinorUndefinedCodes` was updated to remove redundant testing, and the file changed to focus on testing minor undefined currency codes instead.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310923](https://bugs.openjdk.org/browse/JDK-8310923): Refactor Currency tests to use JUnit (**Sub-task** - P4)

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14682/head:pull/14682` \
`$ git checkout pull/14682`

Update a local copy of the PR: \
`$ git checkout pull/14682` \
`$ git pull https://git.openjdk.org/jdk.git pull/14682/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14682`

View PR using the GUI difftool: \
`$ git pr show -t 14682`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14682.diff">https://git.openjdk.org/jdk/pull/14682.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14682#issuecomment-1610333171)

### CODE REVIEW

File: test/jdk/java/util/Currency/MinorUndefinedCodes.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 4512215 4818420 4819436 8310923
+ * @summary Test some minor undefined currencies.
```
Comment (naotoj) [inline test/jdk/java/util/Currency/MinorUndefinedCodes.java:27]: It'd be more readable something like "currencies without minor units." The class name can also be renamed to something like NoMinorUnitCurrenciesTest

File: test/jdk/java/util/Currency/ValidateISO4217.java [hunk 2]
```diff
@@ -28,285 +29,293 @@
...
+
+    // Sets up the following test data:
+    // ISO4217Codes, additionalCodes, testCurrencies, codes
+    @BeforeAll
+    static void setUpTestingData() throws Exception {
+        // These functions laterally setup 'testCurrencies' and 'codes'
+        // at the same time
+        setUpISO4217Codes();
+        setUpAdditionalCodes();
+        finishTestCurrencies();
```
Comment (naotoj) [inline test/jdk/java/util/Currency/ValidateISO4217.java:96]: The method name `'finish` is kind of confusing. I'd explicitly describe what the method does, i.e, setup `other` currencies.

File: test/jdk/java/util/Currency/ValidateISO4217.java [hunk 3]
```diff
@@ -28,285 +29,293 @@
...
+                {"PS", "ILS", "376", "2"},      // Palestinian Territory, Occupied
+                /* Not defined in ISO 4217 list, but added in ISO 3166 country code list */
+                {"JE", "GBP", "826", "2"},      // Jersey
+                {"GG", "GBP", "826", "2"},      // Guernsey
+                {"IM", "GBP", "826", "2"},      // Isle of Man
+                {"BL", "EUR", "978", "2"},      // Saint Barthelemy
+                {"MF", "EUR", "978", "2"},      // Saint Martin
+                /* Defined neither in ISO 4217 nor ISO 3166 list */
+                {"XK", "EUR", "978", "2"},      // Kosovo
+        };
```
Comment (naotoj) [inline test/jdk/java/util/Currency/ValidateISO4217.java:192]: I'd prefer this array to be placed as static in the class, not within the method.

File: test/jdk/java/util/Currency/ValidateISO4217.java [hunk 4]
```diff
@@ -28,285 +29,293 @@
...
+    // The previous set-up method populated most of testCurrencies. This
+    // method finishes populating the list with 'otherCodes'.
+    private static void finishTestCurrencies() {
+        // Codes that are obsolete, do not have related country, extra currency
+        final String otherCodes =
+                "ADP-AFA-ATS-AYM-AZM-BEF-BGL-BOV-BYB-BYR-CHE-CHW-CLF-COU-CUC-CYP-"
+                        + "DEM-EEK-ESP-FIM-FRF-GHC-GRD-GWP-IEP-ITL-LTL-LUF-LVL-MGF-MRO-MTL-MXV-MZM-NLG-"
+                        + "PTE-ROL-RUR-SDD-SIT-SLL-SKK-SRG-STD-TMM-TPE-TRL-VEF-UYI-USN-USS-VEB-VED-"
+                        + "XAG-XAU-XBA-XBB-XBC-XBD-XDR-XFO-XFU-XPD-XPT-XSU-XTS-XUA-XXX-"
+                        + "YUM-ZMK-ZWD-ZWN-ZWR";
```
Comment (naotoj) [inline test/jdk/java/util/Currency/ValidateISO4217.java:214]: Ditto as `extraCodes`

File: test/jdk/java/util/Currency/ValidateISO4217.java [hunk 5]
```diff
@@ -28,285 +29,293 @@
...
+    private static final byte UNDEFINED = 0;
+    private static final byte DEFINED = 1;
+    private static final byte SKIPPED = 2;
+    private static final byte TESTED = 4;
+    private static final int ALPHA_NUM = 26;
+    // An alpha2 code table which maps the status of a country
+    private static final byte[] codes = new byte[ALPHA_NUM * ALPHA_NUM];
+    // Codes derived from ISO4217 golden-data file
+    private static final List<Arguments> ISO4217Codes = new ArrayList<Arguments>();
+    // Additional codes not from the ISO4217 golden-data file
```
Comment (naotoj) [inline test/jdk/java/util/Currency/ValidateISO4217.java:82]: Are these from the golden file?

Comment (justin-curtis-lu) [reply test/jdk/java/util/Currency/ValidateISO4217.java:82]: Unless I am misinterpreting the question, `additionalCodes` is built from the `extraCodes` array, which contains data that is not found in tablea1.txt

File: test/jdk/java/util/Currency/ValidateISO4217.java (hunk 3 above)
Comment (justin-curtis-lu) [reply test/jdk/java/util/Currency/ValidateISO4217.java:192]: Fixed, as well as for `otherCodes`

File: test/jdk/java/util/Currency/ValidateISO4217.java (hunk 2 above)
Comment (justin-curtis-lu) [reply test/jdk/java/util/Currency/ValidateISO4217.java:96]: Renamed it to your suggestion

File: test/jdk/java/util/Currency/CurrencyNameProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.spi.CurrencyNameProvider;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class CurrencyNameProviderTest {
```
[the remark this answers is not in the document]
Comment (justin-curtis-lu) [reply test/jdk/java/util/Currency/CurrencyNameProviderTest.java:43]: I have renamed the file to be more descriptive, but if acronyms are not preferred in file names, I can rename it.

File: test/jdk/java/util/Currency/CurrencyTest.java [hunk 7]
```diff
@@ -38,198 +40,230 @@
...
-                    ", got " + symbol);
-        }
+    private static Stream<Arguments> expectedFractionsProvider() {
+        return Stream.of(
+                Arguments.of("USD", 2), Arguments.of("EUR", 2),
+                Arguments.of("JPY", 0), Arguments.of("XDR", -1),
+                Arguments.of("BHD", 3), Arguments.of("IQD", 3),
+                Arguments.of("JOD", 3), Arguments.of("KWD", 3),
+                Arguments.of("LYD", 3), Arguments.of("OMR", 3),
+                Arguments.of("TND", 3), Arguments.of("TRL", 0), // Turkish Lira
```
[the remark this answers is not in the document]
Comment (justin-curtis-lu) [reply test/jdk/java/util/Currency/CurrencyTest.java:241]: Yes good point, I have clarified the comment

File: test/jdk/java/util/Currency/MinorUndefinedCodes.java (hunk 1 above)
Comment (justin-curtis-lu) [reply test/jdk/java/util/Currency/MinorUndefinedCodes.java:27]: Adjusted the wording throughout the file according to your suggestions

File: test/jdk/java/util/Currency/CurrencyTest.java [hunk 8]
```diff
@@ -238,8 +238,9 @@ private static Stream<Arguments> expectedFractionsProvider() {
                 Arguments.of("BHD", 3), Arguments.of("IQD", 3),
                 Arguments.of("JOD", 3), Arguments.of("KWD", 3),
                 Arguments.of("LYD", 3), Arguments.of("OMR", 3),
-                Arguments.of("TND", 3), Arguments.of("TRL", 0), // Turkish Lira
-                Arguments.of("TRY", 2)
+                // Old and New Turkish Lira
+                Arguments.of("TRL", 0), Arguments.of("TRY", 2),
+                Arguments.of("TND", 3)
```
Comment (naotoj) [inline test/jdk/java/util/Currency/CurrencyTest.java:243]: Probably `TND` moving upward before the comment would be clearer, like:
```
                Arguments.of("LYD", 3), Arguments.of("OMR", 3),
                Arguments.of("TND", 3),

                // Old and New Turkish Lira
                Arguments.of("TRL", 0), Arguments.of("TRY", 2)
```

Comment (justin-curtis-lu) [reply test/jdk/java/util/Currency/CurrencyTest.java]: Adjusted `TND` upwards

**Review by naotoj — APPROVED** [review]
LGTM. Thanks for the refactoring

**Review by LanceAndersen — APPROVED** [review]

**Pull request closed without merging**